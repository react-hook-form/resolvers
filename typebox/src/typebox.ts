import { toNestErrors, validateFieldsNatively } from '@hookform/resolvers';
import { Static, Type } from '@sinclair/typebox';
import { TypeCheck } from '@sinclair/typebox/compiler';
import { Value, type ValueError } from '@sinclair/typebox/value';
import {
  FieldError,
  FieldErrors,
  FieldValues,
  Resolver,
  appendErrors,
} from 'react-hook-form';

function parseErrorSchema(
  _errors: ValueError[],
  validateAllFieldCriteria: boolean,
) {
  const errors: Record<string, FieldError> = {};
  for (; _errors.length; ) {
    const error = _errors[0];
    const { type, message, path } = error;
    const _path = path.substring(1).replace(/\//g, '.');

    if (!errors[_path]) {
      errors[_path] = { message, type: '' + type };
    }

    if (validateAllFieldCriteria) {
      const types = errors[_path].types;
      const messages = types && types['' + type];

      errors[_path] = appendErrors(
        _path,
        validateAllFieldCriteria,
        errors,
        '' + type,
        messages
          ? ([] as string[]).concat(messages as string[], error.message)
          : error.message,
      ) as FieldError;
    }

    _errors.shift();
  }

  return errors;
}

/**
 * Creates a resolver for react-hook-form using Typebox schema validation
 * @param {Schema | TypeCheck<Schema>} schema - The Typebox schema to validate against
 * @param {Object} options - Additional resolver configuration
 * @param {string} [options.mode='async'] - Validation mode
 * @returns {Resolver<Static<Schema>>} A resolver function compatible with react-hook-form
 * @example
 * const schema = Type.Object({
 *   name: Type.String(),
 *   age: Type.Number()
 * });
 *
 * useForm({
 *   resolver: typeboxResolver(schema)
 * });
 */
export function typeboxResolver<
  TFieldValues extends FieldValues,
  Schema extends ReturnType<typeof Type.Object<TFieldValues>>,
>(schema: Schema | TypeCheck<Schema>): Resolver<Static<Schema>> {
  return async (values, _, options) => {
    const errors = Array.from(
      schema instanceof TypeCheck
        ? schema.Errors(values)
        : Value.Errors(schema, values),
    );

    options.shouldUseNativeValidation && validateFieldsNatively({}, options);

    if (!errors.length) {
      return {
        errors: {} as FieldErrors,
        values,
      };
    }

    return {
      values: {},
      errors: toNestErrors(
        parseErrorSchema(
          errors,
          !options.shouldUseNativeValidation && options.criteriaMode === 'all',
        ),
        options,
      ),
    };
  };
}
