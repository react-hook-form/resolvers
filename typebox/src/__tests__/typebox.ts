import { Resolver, SubmitHandler, useForm } from 'react-hook-form';
import { typeboxResolver } from '..';
import { fields, invalidData, schema, validData } from './__fixtures__/data';
import { Type } from '@sinclair/typebox';
import { TypeCompiler } from '@sinclair/typebox/compiler';

const shouldUseNativeValidation = false;

describe('typeboxResolver', () => {
  it('should return a single error from typeboxResolver when validation fails', async () => {
    const result = await typeboxResolver(schema)(invalidData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toMatchSnapshot();
  });

  it('should return all the errors from typeboxResolver when validation fails with `validateAllFieldCriteria` set to true', async () => {
    const result = await typeboxResolver(schema)(invalidData, undefined, {
      fields,
      criteriaMode: 'all',
      shouldUseNativeValidation,
    });

    expect(result).toMatchSnapshot();
  });

  it('should validate with success', async () => {
    const result = await typeboxResolver(schema)(validData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toEqual({ errors: {}, values: validData });
  });

  /**
   * Type inference tests
   */
  it('should correctly infer the output type from a typebox schema', () => {
    const resolver = typeboxResolver(Type.Object({ id: Type.Number() }));

    expectTypeOf(resolver).toEqualTypeOf<
      Resolver<{ id: number }, unknown, { id: number }>
    >();
  });

  it('should correctly infer the output type from a typebox schema with TypeCompiler', () => {
    const typecheck = TypeCompiler.Compile(Type.Object({ id: Type.Number() }));
    const resolver = typeboxResolver(typecheck);

    expectTypeOf(resolver).toEqualTypeOf<
      Resolver<{ id: number }, unknown, { id: number }>
    >();
  });

  it('should correctly infer the output type from a typebox schema using a transform', () => {
    const resolver = typeboxResolver(
      Type.Object({
        id: Type.Transform(Type.Number())
          .Decode((v) => String(v))
          .Encode((v) => Number.parseInt(v)),
      }),
    );

    expectTypeOf(resolver).toEqualTypeOf<
      Resolver<{ id: number }, unknown, { id: string }>
    >();
  });

  it('should correctly infer the output type from a typebox schema for the handleSubmit function in useForm', () => {
    const schema = Type.Object({ id: Type.Number() });

    const form = useForm({
      resolver: typeboxResolver(schema),
      defaultValues: {
        id: 3,
      },
    });

    expectTypeOf(form.watch('id')).toEqualTypeOf<number>();

    expectTypeOf(form.handleSubmit).parameter(0).toEqualTypeOf<
      SubmitHandler<{
        id: number;
      }>
    >();
  });

  it('should correctly infer the output type from a typebox schema with TypeCompiler for the handleSubmit function in useForm', () => {
    const typecheck = TypeCompiler.Compile(Type.Object({ id: Type.Number() }));

    const form = useForm({
      resolver: typeboxResolver(typecheck),
      defaultValues: {
        id: 3,
      },
    });

    expectTypeOf(form.watch('id')).toEqualTypeOf<number>();

    expectTypeOf(form.handleSubmit).parameter(0).toEqualTypeOf<
      SubmitHandler<{
        id: number;
      }>
    >();
  });

  it('should correctly infer the output type from a typebox schema with a transform for the handleSubmit function in useForm', () => {
    const schema = Type.Object({
      id: Type.Transform(Type.Number())
        .Decode((v) => String(v))
        .Encode((v) => Number.parseInt(v)),
    });

    const form = useForm({
      resolver: typeboxResolver(schema),
      defaultValues: {
        id: 3,
      },
    });

    expectTypeOf(form.watch('id')).toEqualTypeOf<number>();

    expectTypeOf(form.handleSubmit).parameter(0).toEqualTypeOf<
      SubmitHandler<{
        id: string;
      }>
    >();
  });
});
