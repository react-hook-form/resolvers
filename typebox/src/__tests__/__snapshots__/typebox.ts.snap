// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typeboxResolver > should return a single error from typeboxResolver when validation fails 1`] = `
{
  "errors": {
    "accessToken": {
      "message": "Expected value of union",
      "ref": undefined,
      "type": "53",
    },
    "birthYear": {
      "message": "Expected number",
      "ref": undefined,
      "type": "29",
    },
    "dateStr": {
      "message": "Expected Date object",
      "ref": undefined,
      "type": "11",
    },
    "email": {
      "message": "Expected string to match format 'email'",
      "ref": {
        "name": "email",
      },
      "type": "48",
    },
    "enabled": {
      "message": "Expected boolean",
      "ref": undefined,
      "type": "10",
    },
    "like": [
      {
        "id": {
          "message": "Expected number",
          "ref": undefined,
          "type": "29",
        },
        "name": {
          "message": "Expected string",
          "ref": undefined,
          "type": "43",
        },
      },
    ],
    "password": {
      "message": "Expected string length greater or equal to 8",
      "ref": {
        "name": "password",
      },
      "type": "44",
    },
    "repeatPassword": {
      "message": "Expected string",
      "ref": undefined,
      "type": "43",
    },
    "tags": {
      "message": "Expected array",
      "ref": undefined,
      "type": "0",
    },
    "username": {
      "message": "Expected string",
      "ref": {
        "name": "username",
      },
      "type": "43",
    },
  },
  "values": {},
}
`;

exports[`typeboxResolver > should return all the errors from typeboxResolver when validation fails with \`validateAllFieldCriteria\` set to true 1`] = `
{
  "errors": {
    "accessToken": {
      "message": "Expected value of union",
      "ref": undefined,
      "type": "53",
      "types": {
        "29": "Expected number",
        "39": "Expected required property",
        "43": "Expected string",
        "53": "Expected value of union",
      },
    },
    "birthYear": {
      "message": "Expected number",
      "ref": undefined,
      "type": "29",
      "types": {
        "29": "Expected number",
      },
    },
    "dateStr": {
      "message": "Expected Date object",
      "ref": undefined,
      "type": "11",
      "types": {
        "11": "Expected Date object",
        "39": "Expected required property",
      },
    },
    "email": {
      "message": "Expected string to match format 'email'",
      "ref": {
        "name": "email",
      },
      "type": "48",
      "types": {
        "48": "Expected string to match format 'email'",
      },
    },
    "enabled": {
      "message": "Expected boolean",
      "ref": undefined,
      "type": "10",
      "types": {
        "10": "Expected boolean",
        "39": "Expected required property",
      },
    },
    "like": [
      {
        "id": {
          "message": "Expected number",
          "ref": undefined,
          "type": "29",
          "types": {
            "29": "Expected number",
          },
        },
        "name": {
          "message": "Expected string",
          "ref": undefined,
          "type": "43",
          "types": {
            "39": "Expected required property",
            "43": "Expected string",
          },
        },
      },
    ],
    "password": {
      "message": "Expected string length greater or equal to 8",
      "ref": {
        "name": "password",
      },
      "type": "44",
      "types": {
        "44": "Expected string length greater or equal to 8",
        "46": "Expected string to match pattern ^(.*[A-Za-z\\\\d].*)$",
      },
    },
    "repeatPassword": {
      "message": "Expected string",
      "ref": undefined,
      "type": "43",
      "types": {
        "39": "Expected required property",
        "43": "Expected string",
      },
    },
    "tags": {
      "message": "Expected array",
      "ref": undefined,
      "type": "0",
      "types": {
        "0": "Expected array",
        "39": "Expected required property",
      },
    },
    "username": {
      "message": "Expected string",
      "ref": {
        "name": "username",
      },
      "type": "43",
      "types": {
        "39": "Expected required property",
        "43": "Expected string",
      },
    },
  },
  "values": {},
}
`;
